### PR TITLE
Codegen filesystem fix

### PR DESCRIPTION
**Pull request by arai713**

### CODE REVIEW

File: codegen/test/rtc/include/rtc/tmp_dir.hpp [hunk 1]
```diff
@@ -2,15 +2,18 @@
 #define GUARD_HOST_TEST_RTC_INCLUDE_RTC_TMP_DIR
 
 #include <string>
-#include <filesystem>
+//#include <filesystem>
```
Comment (pfultz2) [inline codegen/test/rtc/include/rtc/tmp_dir.hpp:5]: Why are we removing `std::filesystem` here?

File: codegen/test/rtc/src/tmp_dir.cpp [hunk 2]
```diff
@@ -50,7 +50,27 @@ std::string tmp_dir::get_tmp_dir_path()
...
+    if(created != 0)
+    {
+        throw std::runtime_error("Directory was not created");
+    }
+}
+
+bool tmp_dir::exists(std::string path)
+{
+    struct stat sb;
+    if(stat(path, &sb) == 0 && !(sb.st_mode & S_IFDIR))
```
Comment (pfultz2) [inline codegen/test/rtc/src/tmp_dir.cpp]: I dont think this will work on windows.

**Comment by pfultz2** [comment]
I am not sure what the motivation is for this change, but there is already work underway to replace this with hiprtc, which removes the need for `std::filesystem`.

File: codegen/test/rtc/include/rtc/tmp_dir.hpp (hunk 1 above)
Comment (illsilin) [reply codegen/test/rtc/include/rtc/tmp_dir.hpp:5]: The problem is that filesystem is not supported on some of the older linux OSs, so this issue has been causing crashes in RHEL/CENTOS dockers and we can't promote CK to staging for some time now.

Comment (pfultz2) [reply codegen/test/rtc/include/rtc/tmp_dir.hpp:5]: Cant you just disable the tests for those older OSs?

Comment (pfultz2) [reply codegen/test/rtc/include/rtc/tmp_dir.hpp:5]: For older OSs, you can use filesystem TS instead. We have this header that defines the `fs` namespace to either `::std::filesystem` or `::std::experimental::filesystem`:

https://github.com/ROCm/AMDMIGraphX/blob/develop/src/include/migraphx/filesystem.hpp

Let me add a comment to the other PR.

Comment (illsilin) [reply codegen/test/rtc/include/rtc/tmp_dir.hpp:5]: yes, we've created another branch where we're using that approach. i just verified that it builds correctly on all OSs.

File: codegen/test/rtc/src/compile_kernel.cpp [hunk 3]
```diff
@@ -70,26 +70,32 @@ kernel compile_kernel(const std::vector<src_file>& srcs, compile_options options
 
     for(const auto& src : srcs)
     {
-        std::filesystem::path full_path   = td.path / src.path;
-        std::filesystem::path parent_path = full_path.parent_path();
-        std::filesystem::create_directories(parent_path);
-        write_string(full_path.string(), src.content);
-        if(src.path.extension().string() == ".cpp")
+        std::string full_path   = td.path + "/" + src.path;
```
Comment (mozga-amd) [inline codegen/test/rtc/src/compile_kernel.cpp:73]: Don't we need to normalize the path by replacing consecutive '/' with a single '/' and removing the trailing '/';? ~ especially for WIN and `mkdir`?

File: codegen/test/rtc/src/compile_kernel.cpp [hunk 4]
```diff
@@ -70,26 +70,32 @@ kernel compile_kernel(const std::vector<src_file>& srcs, compile_options options
...
+        int parent              = full_path.find_last_of('/');
+        std::string parent_path = full_path.substr(0, parent);
+        td.new_dir(parent_path);
+        write_string(full_path, src.content);
+        int ext_pos     = src.path.find_last_of('.');
+        std::string ext = src.path.substr(ext_pos);
+        if(ext == ".cpp")
```
Comment (mozga-amd) [inline codegen/test/rtc/src/compile_kernel.cpp:80]: [To consider:] 
```cpp
int ext_pos = src.path.find_last_of('.');
std::string ext = src.path.substr(ext_pos);
if (ext == ".cpp")
```
replacing with (movie it to functions): 
```cpp
if (hasExtension(".cpp") { } 
/// .... 
inline bool hasExtension(const std::string &mExt) const {
  return endsWith(toLower(src.path), toLower(mExt));
}
// ....
inline bool endsWith(const std::string &mStr, char mChar) noexcept {
  return !mStr.empty() && mStr.back() == mChar;
}
```

File: codegen/test/rtc/src/tmp_dir.cpp [hunk 5]
```diff
@@ -31,18 +35,56 @@ std::string unique_string(const std::string& prefix)
...
+    if(created != 0)
+    {
+        throw std::runtime_error("Directory was not created");
+    }
+}
+
+bool tmp_dir::exists(std::string check_path)
+{
+    struct stat sb;
+    if(stat(check_path.c_str(), &sb) == 0 && !(sb.st_mode & S_IFDIR))
```
Comment (mozga-amd) [inline codegen/test/rtc/src/tmp_dir.cpp:68]: Could you please use bitwise `AND-ing` directly with `sb.st_mode`:  like `(sb.st_mode & S_IFMT) == S_IFDIR; `  ~ to extract the file type code, and compared to the appropriate constant.

File: codegen/test/rtc/src/tmp_dir.cpp [hunk 6]
```diff
@@ -31,18 +35,56 @@ std::string unique_string(const std::string& prefix)
...
+    {
+        return "/tmp";
+    }
+    return tmp_dir_path;
+}
+
+// TODO: finish this method
+void tmp_dir::new_dir(std::string dir_path)
+{
+    int created = mkdir(dir_path.c_str(), 0755);
```
Comment (mozga-amd) [inline codegen/test/rtc/src/tmp_dir.cpp:58]: If we want to create a temporary file, we can use `std::filesystem::temp_directory_path()`, which calls `mkdir`. Alternatively, we can use `mkdtemp` from POSIX:
```cpp
std::string pattern = "tmep/nnn-xxxx";
if (mkdtemp(pattern.data()) == NULL) { }
```

**Comment by illsilin** [comment]
This PR is no longer necessary, it's been replaced by #1509 .

**Pull request closed without merging**